### PR TITLE
enh: support visiting module in pass array by data

### LINKED ISSUE #3728: ArraySize node is not replaced for function contained in module
```fortran
module xx
contains
function var_2_iint8_dp(x, dim, mask, corrected) result(res)
use iso_fortran_env, only: dp => real64
integer, intent(in) :: x(:,:)
integer, intent(in) :: dim
logical, intent(in), optional :: mask
logical, intent(in), optional :: corrected
real(dp) :: res
print *, sum(real(x, dp), dim)
end function var_2_iint8_dp
end module
```

> cat pass_22_array_dim_intrinsics_update.clj

```clojure
[((ArraySize
    (Cast
        (Var 3 x)
        IntegerToReal
        (Array
            (Real 8)
            [(()
            ())
            (()
            ())]
            DescriptorArray
        )
        ()
    )
    (IntegerConstant 1 (Integer 4))
    (Integer 4)
    ()
))
((ArraySize
    (Cast
        (Var 3 x)
        IntegerToReal
        (Array
            (Real 8)
            [(()
            ())
            (()
            ())]
            DescriptorArray
        )
        ()
    )
    (IntegerConstant 2 (Integer 4))
    (Integer 4)
    ()
))
((IntegerCompare
    (IntegerConstant 1 (Integer 4))
    Lt
    (Var 3 dim)
    (Logical 4)
    ()
))]
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #3728.

With this we get `stdlib_stats_var` compiled to LLVM.

### CODE REVIEW

**Comment by Pranavchiku** [comment]
CI Failure:

```console
Test project /home/runner/work/lfortran/lfortran
    Start 1: test_cwrapper
1/3 Test #1: test_cwrapper ....................   Passed    0.00 sec
    Start 2: test_stacktrace
2/3 Test #2: test_stacktrace ..................   Passed    0.00 sec
    Start 3: test_lfortran
3/3 Test #3: test_lfortran ....................***Exception: SegFault  0.11 sec
[doctest] doctest version is "2.4.8"
```

I am not sure why it fails

**Comment by Pranavchiku** [comment]
CI should pass now.

**Comment by Pranavchiku** [comment]
This is weird, it works well for me now:

```console
% ctest
Test project /Users/pranavchiku/repos/lfortran
    Start 1: test_cwrapper
1/3 Test #1: test_cwrapper ....................   Passed    0.02 sec
    Start 2: test_stacktrace
2/3 Test #2: test_stacktrace ..................   Passed    0.01 sec
    Start 3: test_lfortran
3/3 Test #3: test_lfortran ....................   Passed    0.30 sec

100% tests passed, 0 tests failed out of 3

Total Test time (real) =   0.35 sec
```

**Comment by certik** [comment]
The failure is:
```
/home/runner/work/lfortran/lfortran/lfortran-0.34.0-76-gc1f6546c3/src/lfortran/tests/test_llvm.cpp:360:
TEST CASE:  ASR -> LLVM 1

/home/runner/work/lfortran/lfortran/lfortran-0.34.0-76-gc1f6546c3/src/lfortran/tests/test_llvm.cpp:360: FATAL ERROR: test case CRASHED: SIGSEGV - Segmentation violation signal
```

**Comment by certik** [comment]
It fails for me too, here is the stacktrace:
```console
(lf) lfortran/lfortran(gh3728) $ lldb ./src/lfortran/tests/test_lfortran
(lldb) target create "./src/lfortran/tests/test_lfortran"
Current executable set to '/Users/ondrej/repos/lfortran/lfortran/src/lfortran/tests/test_lfortran' (arm64).
(lldb) run
Process 14926 launched: '/Users/ondrej/repos/lfortran/lfortran/src/lfortran/tests/test_lfortran' (arm64)
[doctest] doctest version is "2.4.8"
[doctest] run with "--help" for options
Process 14926 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0xfffffffffffff00c)
    frame #0: 0x00000001018860d0 test_lfortran`bool LCompilers::ASR::is_a<LCompilers::ASR::Class_t, LCompilers::ASR::ttype_t>(x=0xfffffffffffff000) at asr.h:32:31
   29  	template <class T, class U>
   30  	inline bool is_a(const U &x)
   31  	{
-> 32  	    return T::class_type == x.type;
   33  	}
   34  	
   35  	// Cast one level down
Target 0: (test_lfortran) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0xfffffffffffff00c)
  * frame #0: 0x00000001018860d0 test_lfortran`bool LCompilers::ASR::is_a<LCompilers::ASR::Class_t, LCompilers::ASR::ttype_t>(x=0xfffffffffffff000) at asr.h:32:31
    frame #1: 0x00000001016a5fa0 test_lfortran`void LCompilers::RemoveArrayByDescriptorProceduresVisitor::visit_Unit<LCompilers::ASR::TranslationUnit_t>(this=0x000000016fdfde60, x=0x000000010782dcd8) at pass_array_by_data.cpp:649:54
    frame #2: 0x000000010166e1d4 test_lfortran`LCompilers::RemoveArrayByDescriptorProceduresVisitor::visit_TranslationUnit(this=0x000000016fdfde60, x=0x000000010782dcd8) at pass_array_by_data.cpp:678:13
    frame #3: 0x000000010166dca8 test_lfortran`LCompilers::pass_array_by_data(al=0x000000016fdfedb8, unit=0x000000010782dcd8, (null)=0x000000016fdfe988) at pass_array_by_data.cpp:713:7
    frame #4: 0x0000000101cecd08 test_lfortran`LCompilers::PassManager::apply_passes(this=0x000000016fdfeae0, al=0x000000016fdfedb8, asr=0x000000010782dcd8, passes=size=30, pass_options=0x000000016fdfe988, diagnostics=0x000000016fdfed90) at pass_manager.h:167:17
    frame #5: 0x0000000101cac0f8 test_lfortran`LCompilers::PassManager::apply_passes(this=0x000000016fdfeae0, al=0x000000016fdfedb8, asr=0x000000010782dcd8, pass_options=0x000000016fdfe988, diagnostics=0x000000016fdfed90) at pass_manager.h:311:21
    frame #6: 0x0000000101caba40 test_lfortran`LCompilers::asr_to_llvm(asr=0x000000010782dcd8, diagnostics=0x000000016fdfed90, context=0x00006000000042e0, al=0x000000016fdfedb8, pass_manager=0x000000016fdfeae0, co=0x000000016fdfe970, run_fn="f", infile="") at asr_to_llvm.cpp:9792:18
    frame #7: 0x000000010009f35c test_lfortran`DOCTEST_ANON_FUNC_18() at test_llvm.cpp:390:15
    frame #8: 0x00000001007a2360 test_lfortran`doctest::Context::run(this=0x000000016fdff5e8) at doctest.h:6727:21
    frame #9: 0x00000001007a4210 test_lfortran`main(argc=1, argv=0x000000016fdff778) at doctest.h:6805:71
    frame #10: 0x00000001070990f4 dyld`start + 520
```

File: src/libasr/pass/pass_array_by_data.cpp [hunk 1]
```diff
@@ -644,11 +644,29 @@ class RemoveArrayByDescriptorProceduresVisitor : public PassUtils::PassVisitor<R
             std::vector<std::string> to_be_erased;
 
             for( auto& item: current_scope->get_scope() ) {
-                if( v.proc2newproc.find(item.second) != v.proc2newproc.end() &&
-                    not_to_be_erased.find(item.second) == not_to_be_erased.end() ) {
+                if (ASR::is_a<ASR::Function_t>(*item.second)) {
+                    ASR::FunctionType_t* func_type = ASRUtils::get_FunctionType(item.second);
+                    if (func_type->m_arg_types[0] && ASR::is_a<ASR::Class_t>(*func_type->m_arg_types[0])) {
```
Comment (certik) [inline src/libasr/pass/pass_array_by_data.cpp:649]: You are accessing a dangling pointer, you have to do it like this:
```suggestion
                    if (func_type->n_arg_types >= 1 && ASR::is_a<ASR::Class_t>(*func_type->m_arg_types[0])) {
```

**Review by certik — APPROVED** [review]
I think this is fine.

**Comment by Pranavchiku** [comment]
Aah, thanks for the fix! I’ll look into stdlib failure and get it fixed soon.

**Comment by certik** [comment]
The stdlib failure is:
```
ASR verify pass error: ExternalSymbol::m_original_name ('gauss_legendre_fp64') + scope_names not found in a module 'stdlib_quadrature'
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
LCompilersException: Verify failed in the pass: pass_array_by_data
```

**Comment by Pranavchiku** [comment]
MRE:

```fortran
module xx
use iso_fortran_env, only: dp => real64

interface gauss_legendre
procedure :: gauss_legendre_fp64
end interface gauss_legendre

contains
subroutine gauss_legendre_fp64(x)
    real(dp), intent(out) :: x(:)
end subroutine
end module

subroutine test_gauss_analytic()
    use stdlib_kinds, only: dp
    use xx, only: gauss_legendre
    block
        real(dp), dimension(5) :: x
        call gauss_legendre(x)
    end block
end subroutine
```

**Comment by Pranavchiku** [comment]
I have CI changes ready, will open PR once this gets merged.

**Comment by certik** [comment]
Perfect, thanks!

**Pull request merged**